shared/js/participantStore.js

```diff
@@ -0,0 +1,96 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this file,
+ * You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+var loop = loop || {};
+loop.store = loop.store || {};
+
+loop.store.ParticipantStore = (function() {
+  "use strict";
+
+  const PARTICIPANT_SCHEMA = {
+    name: "",
+    isHere: false,
+    pingTime: null
+  };
+
+  /**
+   * Participant store.
+   *
+   * @param {loop.Dispatcher} dispatcher  The dispatcher for dispatching actions
+   *                                      and registering to consume actions.
+   */
+  var ParticipantStore = loop.store.createStore({
+    getInitialStoreState: function() {
+      return {
+        participants: new Map()
+      };
+    },
+
+    actions: [
+      "updatedParticipant",
+      "updatedPresence"
+    ],
+
+    /**
+     * Checks if the room is empty or has participants.
+     */
+    _hasParticipants: function() {
+      return this._storeState.participants.size > 0;
+    },
+
+    updatedParticipant: function(actionData) {
+      this._updateParticipantData(actionData.userId, {
+        name: actionData.name
+      });
+    },
+
+    updatedPresence: function(actionData) {
+      this._updateParticipantData(actionData.userId, {
+        isHere: actionData.isHere,
+        pingTime: actionData.pingTime
+      });
+    },
+
+    _updateParticipantData: function(userId, updatedData) {
+      let updatedParticipant = this._storeState.participants.get(userId);
+      if (!updatedParticipant) {
+        updatedParticipant = _.extend({}, PARTICIPANT_SCHEMA);
+      }
+
+      for (let key in updatedData) {
+        if (updatedParticipant.hasOwnProperty(key)) {
+          updatedParticipant[key] = updatedData[key];
+        }
+      }
+
+      this.setStoreState({
+        participants: this._storeState.participants.set(userId, updatedParticipant)
+      });
+    },
+
+    /*
+     * Gets the online participants in the room taking into account
+     * unexpected disconnects/leaves
+     */
+    getOnlineParticipants: function() {
+      let participants = this._storeState.participants;
+      let onlineParticipants = [];
+      participants.forEach(function(participant) {
+        let pingTime = new Date(participant.pingTime);
+        let maxPingTimeAllowed = new Date();
+        // XXX akita: Max ping time allowed is not defined yet (1 minute by default)
+        // If you change the timeout, don't forget to update the test
+        maxPingTimeAllowed.setMinutes(maxPingTimeAllowed.getMinutes() - 1);
+        if (participant.isHere && pingTime.getTime() >= maxPingTimeAllowed.getTime()) {
```

Comment:
I think it makes sense to have the UpdatedParticipant action to have `pingedAgo` instead of `pingTime` to avoid needing to expose action handlers to `DataDriver` to get the server time. Within `ParticipantStore` on receiving a `pingedAgo`, the value stored into the `Map` could be a single value of `localPingTime: Date.now() - pingedAgo` then in the dynamic calls to `getOnlineParticipants` calculate `Date.now() - localPingTime > 1 minute`